sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```diff
@@ -0,0 +1,317 @@
+      }
+    }
+
+    @Override
+    public T getCurrent() throws NoSuchElementException {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+  }
+
+  @Override
+  public Reader<T> createReader(String query, Class<T> entity) throws IOException {
+    return new CassandraReaderImpl<>(query, entity);
+  }
+
+  @Override
+  public long getEstimatedSizeBytes() {
+    try (Cluster cluster = getCluster()) {
+      List<TokenRange> tokenRanges = getTokenRanges(cluster, keyspace, table);
+      long size = 0L;
+      for (TokenRange tokenRange : tokenRanges) {
+        size += tokenRange.meanPartitionSize * tokenRange.partitionCount;
+      }
+      return Math.round(size / getRingFraction(cluster, tokenRanges));
+    }
+  }
+
+  @Override
+  public List<BoundedSource<T>> splitIntoBundles(
+      CassandraIO.Read<T> spec,
+      long desiredBundleSizeBytes) {
+    long numSplits = 1;
+    List<BoundedSource<T>> sourceList = new ArrayList<>();
+    try {
+      if (desiredBundleSizeBytes > 0) {
+        numSplits = getEstimatedSizeBytes() / desiredBundleSizeBytes;
+      }
+    } catch (Exception e) {
+      LOG.warn("EstimatedSizeBytes / DesiredBundleSizeBytes failed, fallback to an unique split",
+          e);
+      numSplits = 1;
+    }
+    if (numSplits <= 0) {
+      LOG.warn("Number of splits is less than 0 ({}), fallback to 1", numSplits);
+      numSplits = 1;
+    }
+
+    LOG.debug("Number of splits is {}", numSplits);
+
+    if (numSplits == 1) {
+      sourceList.add(new CassandraIO.CassandraSource<T>(spec, null));
+      return sourceList;
+    }
+
+    Partitioner partitioner = new Partitioner(getCluster());
+
+    BigInteger startRange = partitioner.minToken;
+    BigInteger endRange = partitioner.maxToken;
+    BigInteger startToken, endToken;
+
+    endToken = startRange;
+    BigInteger incrementValue =
+        (endRange.subtract(startRange)).divide(new BigInteger(String.valueOf(numSplits)));
+    String splitQuery;
+    for (int splitCount = 1; splitCount <= numSplits; splitCount++) {
+      startToken = endToken;
+      endToken = startToken.add(incrementValue);
+      if (splitCount == numSplits) {
+        endToken = endRange;
+      }
+      splitQuery =
+          QueryBuilder.select()
+              .from(keyspace, table)
+              .where(QueryBuilder.gte("token($pk)", startToken))
+              .and(QueryBuilder.lt("token($pk)", endToken))
+              .toString();
+      sourceList.add(new CassandraIO.CassandraSource(spec, splitQuery));
+    }
+    return sourceList;
+  }
+
+  /**
+   * Get a Cassandra cluster using hosts and port.
+   */
+  private Cluster getCluster() {
+    return Cluster.builder()
+        .addContactPoints(hosts.toArray(new String[0]))
+        .withPort(port)
+        .build();
+  }
+
+  /**
+   * Gets the list of token ranges that a table occupies on a give Cassandra node.
+   */
+  //compatible Cassandra > 2.1.5
```

Comment:
Put into javadoc, and into javadoc of the main class - it's important for a user to know that this only works with Cassandra past 2.1.5.